### ISSUE TITLE
[HOLD for payment 2022-12-23] [$2000] [Bug] Keyboard being open/closed changes position of ToS on certain mobile platforms - reported by @gadhiyamanan

### ISSUE THREAD

**Issue by mvtglobally** [Expensify/App#11277]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open login page on a fresh install OR after clearing app data
2. Enter email/phone number.
3. Open the touch screen keyboard by tapping on the login text input. Notice how the Terms and Licenses don't shift upwards to avoid the keyboard.
4. Press "Continue" to get to the password form.
5. Repeat step 3 with the password input. Notice how the Terms and Licenses doesn't shift upwards to avoid the keyboard.

## Expected Result:
- The Terms and Licenses should shift upwards to avoid the keyboard on both the login and password forms.

## Actual result:
- The page becomes scrollable, and you can scroll the login/password input out of view or out of view
- The Terms and Licenses are not visible

## Workaround:
None 

## Platform:

Where is this issue occurring?

- iOS Safari
- iOS Native
- Android Chrome

**Version Number:**  1.2.4-0
**Reproducible in staging?:** need repro
**Reproducible in production?:** need repro
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation 

<details>
<summary>iOS - Both forms are broken</summary>

https://user-images.githubusercontent.com/31285285/206932042-26b677cf-3b09-464d-8f0c-0e24fbf14bbf.mov

https://user-images.githubusercontent.com/31285285/206932045-21522b6c-f073-4a5f-b3c3-50f5b6d5f7ce.mov
</details>

<details>
<summary>Mobile Safari - Both forms are broken</summary>

https://user-images.githubusercontent.com/31285285/206932114-476668ac-3dc9-4e39-9f53-29bec36c7f1f.mov

https://user-images.githubusercontent.com/31285285/206932117-b3401cd8-10ea-4d10-a913-f6b850f5b136.mov
</details>

<details>
<summary>Mobile Chrome - Interestingly, it appears to work on the login form, but not the password form</summary>

https://user-images.githubusercontent.com/31285285/206932196-a47f0f5c-79bc-48ac-9629-1a958bda5f42.mov

https://user-images.githubusercontent.com/31285285/206932189-1ea13dd0-d035-4326-b8b0-00611378ba7f.mov
</details>

**Expensify/Expensify Issue URL:** 
**Issue reported by:** @gadhiyamanan
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1661843853973199

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by JmillsExpensify** [comment]
I'm on iOS, so I'm unable to reproduce. That said, can you clarify exactly what you mean by "instruction goes up?" I'm not following what the issue is here. Thank you!

**Comment by gadhiyamanan** [comment]
@JmillsExpensify here instruction means **By logging in, you agree to the...** , when you click on the phone field there is a large gap between the continue button and the instruction. after typing any word, space will remove

it's reproduced on android

**Comment by JmillsExpensify** [comment]
Ok thanks. Still unable to reproduce since I don't have an android device. That said, I have a suspicion that this is yet another symptom of our keyboard implementation. @roryabraham do you mind confirming if you think my hunch is right? If so, I'll update the issue title, put this issue on hold, and add it to the main tracker.

**Comment by roryabraham** [comment]
Okay, I was able to reproduce this, going to update the OP to add some additional details then I agree we should add it to the keyboard tracking issue (cc @tgolen)

**Comment by JmillsExpensify** [comment]
Thanks @roryabraham! I'll handle adding to the keyboard tracking issue.

**Comment by roryabraham** [comment]
Added clarification and consistent reproduction steps, removed the `Needs Reproduction` label.

**Comment by roryabraham** [comment]
Also for transparency I was testing with a physical Pixel 4a

**Comment by JmillsExpensify** [comment]
Thanks @roryabraham.

**Comment by JmillsExpensify** [comment]
Still on hold.

**Comment by JmillsExpensify** [comment]
Still on hold.

**Comment by JmillsExpensify** [comment]
Still on hold for the related keyboard initiative.

**Comment by tgolen** [comment]
This is no longer on hold and can have proposals created for it with the new code.

**Comment by JmillsExpensify** [comment]
Alright cool. Given how many issues we have related to the keyboard, I'm going to start this issue at $500.

**Comment by JmillsExpensify** [comment]
Upwork job is here: https://www.upwork.com/jobs/~0164d0286c2126643f. All proposals welcome!

**Comment by salahawk** [comment]
> Upwork job is here: https://www.upwork.com/jobs/~0164d0286c2126643f. All proposals welcome!

I just checked this job description on Upwork carefully and it seems like we should change the layout of that page.

**Comment by saadyousaf1997** [comment]
> Upwork job is here: https://www.upwork.com/jobs/~0164d0286c2126643f. All proposals welcome!

I have read the issue, I think we need to add keyboardDidShow event for android, When the event is emitted we will set the height of the layout based on the keyboard opened or closed.

**Comment by JmillsExpensify** [comment]
Make sure you're following our guidelines for [posting proposals](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#posting-ideas). Neither of those post qualify as a proposal that we can evaluate and approve.

**Comment by ivanmiklec** [comment]
Proposal: `KeyboardAvoidingView` does not work for all android keyboards, and interacts differently with iOS and Android. Since behavior is only wrong on Android, I would set the `behavior` specifically for Android.  

Also, looking at your  `android/app/src/main/AndroidManifest.xml`, option `android:windowSoftInputMode` is set to
`adjustResize`, meaning whenever the keyboard is opened, components are resized to fit everything on the screen. That behavior can be a bit flaky, and setting `adjustNothing` and implementing custom behavior could be more beneficial in the long run.

**Comment by sobitneupane** [comment]
@ivanmiklec If you haven’t already, check out [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md).

Please check out Expected Results in [issue description](https://github.com/Expensify/App/issues/11277#issue-1385474423).

> whenever the keyboard is opened, components are resized to fit everything on the screen. 

We are using this feature in many screens.

> implementing custom behavior could be more beneficial in the long run.

What are the changes you are proposing? 

I believe there should be a simple solution to this issue.

**Comment by JmillsExpensify** [comment]
Still waiting for proposals.

**Comment by tgolen** [comment]
Waiting for proposals

**Comment by JmillsExpensify** [comment]
We've hit the 7 day mark since the last price increase, so I'm going to increase the [Upwork job](https://www.upwork.com/jobs/~0164d0286c2126643f) to $1,000. We're still looking for a solid proposal!

**Comment by sobitneupane** [comment]
Waiting for proposal.

**Comment by jayeshmangwani** [comment]
@gadhiyamanan can you please check this issue one more time from the latest main branch? for me, this issue happening in the play store build but not in the latest main branch

**Comment by JmillsExpensify** [comment]
Unfortunately don't have a way to test Android at the moment, so I'll have to rely on someone else jumping in to help confirm if this is still reproducible.

**Comment by sobitneupane** [comment]
Was able to reproduce.


https://user-images.githubusercontent.com/25876548/201616289-71883ebd-894c-4b76-918d-fdd11b7f3840.mov

**Comment by JmillsExpensify** [comment]
Thanks @sobitneupane! Question how did you feel about the testing steps split up the way they currently are, essentially the two expected/actual results? Was that clear for you?

**Comment by JmillsExpensify** [comment]
Then separately, it's been a week, so time to update the price of this issue. Now we're at $2,000. Upwork job is here: https://www.upwork.com/jobs/~0164d0286c2126643f.

**Comment by sobitneupane** [comment]
> Question how did you feel about the testing steps split up the way they currently are, essentially the two expected/actual results? Was that clear for you?

Yes. it was clear.

But reproduction steps for "Action Performed (part 1):" was quite different for me. I reproduced it with following testing steps:

Action Performed (part 1):
1. Open login page on a fresh install OR after clearing app data
2. Enter email/phone number. > Continue
3. Close the keyboard and press on Go back to go to login page

**Comment by JmillsExpensify** [comment]
Ok, cool. Thanks for confirming! I will update part 1 to make sure it's in sync with the steps/results of our latest test.

**Comment by JmillsExpensify** [comment]
Still waiting on proposals. Ideally we have Margelo or an internal engineer take this over since we're still struggling to get proposals.

**Comment by jasperhuangg** [comment]
Gonna look into this one

**Comment by railway17** [comment]
`Help wanted` label should be removed and not waiting for external proposal?

**Comment by s77rt** [comment]
is this still reproducible?

**Comment by tgolen** [comment]
I'm going to put this on a temporary HOLD because I made some refactorings in https://github.com/Expensify/App/pull/12848 which will change any proposals in this issue. Once my PR is merged, then the new code can be re-evaluated. I believe the issue still remains in Android due to differences in how the screen height changes in regards to the keyboard being opened (Android changes the size of the page, iOS just moves the contents out of view).

I personally think either behavior is fine though and we don't need it to be 100% the same in both iOS and Android, as long as the user can access the link for TOS, and the entire login form is in view, then I think that's all that matters.

**Comment by jasperhuangg** [comment]
@s77rt here's it still happening on iOS Chrome, the ToS text and the Expensify logo should move upwards to avoid the keyboard and stay in view:

#### Staging
https://user-images.githubusercontent.com/31285285/203090559-f5c060a1-fc81-4cce-8a5b-0f5c266e2e65.MP4

#### Prod
https://user-images.githubusercontent.com/31285285/203090511-504ab4c6-75ad-4bc5-9413-5bb603fc25a7.MP4

**Comment by jasperhuangg** [comment]
@tgolen Thanks for letting me know! I actually have a branch locally that fixes the issue which is why I removed the `Help Wanted` label. The diff for my branch is similar to what you have for https://github.com/Expensify/App/pull/12848 😆 so they would totally conflict and I agree we should HOLD

I just checked out your branch and this issue seems to still be reproducible there, so this issue should definitely remain open.

**Comment by jasperhuangg** [comment]
@tgolen Thanks for letting me know! I actually have a branch locally that fixes the issue which is why I removed the `Help Wanted` label. The diff for my branch touches the same files as https://github.com/Expensify/App/pull/12848 😆 so they would totally conflict and I agree we should HOLD

I just checked out your branch and this issue seems to still be reproducible there, so this issue should definitely remain open.

**Comment by JmillsExpensify** [comment]
Still waiting on the linked PR.

**Comment by JmillsExpensify** [comment]
Still waiting for the linked PR to be merged.

**Comment by tgolen** [comment]
The linked PR is looking to be merged this week 🤞

**Comment by JmillsExpensify** [comment]
Looks like that linked PR is in the final stages of review.

**Comment by JmillsExpensify** [comment]
Linked PR should hopefully be merged today.

**Comment by jasperhuangg** [comment]
For ease of access @JmillsExpensify is referring to https://github.com/Expensify/App/pull/12848 above!

Looks like it was just merged 🚀 excited to get working on this

**Comment by tgolen** [comment]
I've removed the HOLD on this and I'm unassigning myself since it sounds like Jasper is excited to do this 👍

**Comment by JmillsExpensify** [comment]
Woo! Thanks for driving this forward.

**Comment by jasperhuangg** [comment]
Finally done with the testing on https://github.com/Expensify/App/issues/12573, taking a look into this one now

**Comment by jasperhuangg** [comment]
Quick update on this one, thanks for your patience, was dealing with emulator issues last week. Just wanted to give a current state of the union:

<details>
<summary> Android - This issue seems to have been fixed in Tim's PR </summary>

https://user-images.githubusercontent.com/31285285/206932009-460df5ae-8642-44e5-8167-27dcd5771be7.mov

https://user-images.githubusercontent.com/31285285/206932016-27b62716-f628-46ef-b860-057729cb4692.mov
</details>


<details>
<summary>iOS - KAV is broken</summary>

https://user-images.githubusercontent.com/31285285/206932042-26b677cf-3b09-464d-8f0c-0e24fbf14bbf.mov

https://user-images.githubusercontent.com/31285285/206932045-21522b6c-f073-4a5f-b3c3-50f5b6d5f7ce.mov
</details>

<details>
<summary>Mobile Safari - KAV is broken</summary>

https://user-images.githubusercontent.com/31285285/206932114-476668ac-3dc9-4e39-9f53-29bec36c7f1f.mov

https://user-images.githubusercontent.com/31285285/206932117-b3401cd8-10ea-4d10-a913-f6b850f5b136.mov
</details>

<details>
<summary>Mobile Chrome - Interestingly, it appears to work on the login form, but not the password form</summary>

https://user-images.githubusercontent.com/31285285/206932196-a47f0f5c-79bc-48ac-9629-1a958bda5f42.mov

https://user-images.githubusercontent.com/31285285/206932189-1ea13dd0-d035-4326-b8b0-00611378ba7f.mov
</details>

**Comment by jasperhuangg** [comment]
I'm realizing now that the `withKeyboardState` HOC doesn't actually work on Mobile Safari. When I log `props.isShown` (which is passed down to children from `withKeyboardState`), it's always false, regardless of whether the keyboard is open or not.

I'm considering using this state variable to create some spacing underneath the ToS, similar to what Tim did in his PR.

**Comment by jasperhuangg** [comment]
Reaching out for some help on this one since I've exhausted all my possible ideas. Thread [here](https://expensify.slack.com/archives/C03TQ48KC/p1670803198625219)!

**Comment by tgolen** [comment]
@jasperhuangg I think it's a good sense that you picked up on the problem with `props.isShown` for mobile web. I recently discussed this over in another issue, you can see my main comment [here](https://github.com/Expensify/App/issues/11094#issuecomment-1324296237), but you might need to read some of the context around it.

There was a fix for mobileWeb in the initial proposal for that issue:

> checking `(prevProps.windowHeight < this.props.windowHeight)` in componendDidUpdate in `BaseOptionsSelector.js`

I felt like it was a hack to leave it there and suggested moving that fix into `withKeyboardState` so that `withKeyboardState` would have better support for mobileWeb. The contributor did not want to do that, and so we agreed to just not implement the "hack" in `BaseOptionsSelector.js` for that issue.

I still think that adding mobileWeb support into `withKeyboardState` is the best solution, so maybe you can explore that more? I think we need to do a little bit more research on it, but the idea is to use screen size detection on mobileWeb to try and guess when the software keyboard might be active. So, something like this:

- If the screen matches `isSmallScreenWidth` which comes from `withWindowDimensions`
- then listen for screen resize event
- if the height goes from taller to shorter, then assume the keyboard just opened
- if the height goes from shorter to taller, then assume the keyboard just closed

**Comment by tgolen** [comment]
To test out my theory, I went to `SigninPageContent.js` and placed this log at the top of the render method:

```
console.log('!!!', props.isShown, props.windowHeight);
```

What I found is that on iOS, when the keyboard is shown/dismissed, the `props.isShown` works, but the problem is that the window height does not ever change. So, that's not going to work.

**Comment by tgolen** [comment]
I also attempted to use the reanimated hook that was originally the thought behind the proposal in https://github.com/Expensify/App/issues/10273 but I also found it does not work for mobile safari.

So, basically, my conclusion is that for mobile safari, we just can't really provide good keyboard support and the best fallback is letting the view be scrollable, and let inputs be blurrable, so someone can still get to the content.

I also found that with iOS native, adding a KAV to the `SigninPageContent.js` in the right spot does help slightly. You can see in this screenshot how the terms are starting to be visible. Sorry for the colors. It really helps to identify which components are stretching and flexing properly. It's interesting that even with a KAV, the height of the window still isn't adjusting properly for iOS.

![image](https://user-images.githubusercontent.com/1228807/207128581-0a56e82a-8287-4e1b-b46c-263df38a1e48.png)

**Comment by jasperhuangg** [comment]
@tgolen Thanks for taking the time to help out here. Just updated the OP like you asked.

> I also found that with iOS native, adding a KAV to the SigninPageContent.js in the right spot does help slightly. You can see in this screenshot how the terms are starting to be visible.

Wow interesting, when you say "right spot" what do you mean? It looks like the margin above the Terms and Licenses matches the height of the cutoff portion below. And no worries about the colors haha, my simulator is literally a rainbow when working on these types of issues too 😆 

> What I found is that on iOS, when the keyboard is shown/dismissed, the props.isShown works, but the problem is that the window height does not ever change. So, that's not going to work.

You mentioned updating `withKeyboardState` to work for iOS mobile using the window dimensions approach. I just tried the same litmus test you used, and it appears that the window height also doesn't change on Mobile Safari. That makes sense, considering how the entire view is still scrollable. 

> So, basically, my conclusion is that for mobile safari, we just can't really provide good keyboard support and the best fallback is letting the view be scrollable, and let inputs be blurrable, so someone can still get to the content.

Gotcha, so to clarify, for now the main focus should be trying to get it to work on iOS native and on the password page on Mobile Chrome?

Since this issue is so platform specific, my question is this: ideally, we'd want an over-arching solution that can work for all platforms, but at this point, would it be wrong to consider splitting the login/password forms into different components for each platform? That's where my brain immediately goes

**Comment by tgolen** [comment]
> Wow interesting, when you say "right spot" what do you mean?

You can see [here](https://github.com/Expensify/App/pull/13529/files#diff-dd48053a4b4ca8a948af29c248e8fee97c8d962017338228d8bec523842df230R52) where I've added it. If you don't put it in the right spot in the view hierarchy, then you get a bunch of funky layout issues (like everything smooshed into 50 pixels at the top of the page, or the page not spanning the full vertical height).

>  ideally, we'd want an over-arching solution that can work for all platforms, but at this point, would it be wrong to consider splitting the login/password forms into different components for each platform?

I appreciate the thought, but I think that would be one of the worst solutions 😬 so I would want to avoid that at all costs.

I also find that `isShown` isn't working for mobile Chrome too (eg. it's always false), which doesn't surprise me a whole lot.

I think all that we should focus on right now are:

- Getting the terms to display above the keyboard for both iOS and Android native on both login and password forms

I've updated my PR to achieve this, but some of the styles need cleaned up. I accomplish it by adding some margin to the bottom of the terms container when the keyboard is open. However, this has several drawbacks:

1. It makes the animation a little janky because the margin isn't added until the keyboard animation ends
2. Android doesn't show the extra margin as nicely as iOS does and leaves extra space below the terms (this could be fixed by only applying the 60px of margin to iOS)

<details>
<summary>iOS - shows the janky animation</summary>

https://user-images.githubusercontent.com/1228807/207158611-2251df1e-b7bf-43ea-a0de-9314bd001d97.mp4


</details>

<details>
<summary>Android - shows the extra margin at the bottom</summary>

https://user-images.githubusercontent.com/1228807/207158834-4904b217-2949-41c2-ad24-51b8c743710f.mp4


</details>

**Comment by jasperhuangg** [comment]
> I appreciate the thought, but I think that would be one of the worst solutions 😬 so I would want to avoid that at all costs.

hehe yup, my brain went there, but it just felt soooo wrong 😆 

> I also find that isShown isn't working for mobile Chrome too (eg. it's always false), which doesn't surprise me a whole lot.

Yup, the `withKeyboardState` HOC doesn't work at all with any of the mobile web platforms, because React Native's `Keyboard` library it uses only works on native platforms :/

> I've updated my PR to achieve this, but some of the styles need cleaned up. I accomplish it by adding some margin to the bottom of the terms container when the keyboard is open. However, this has several drawbacks:

Gotcha, thanks for clarifying. Sorry I haven't had the chance to dive into your PR, but my next steps are clear. I'll help you look into those styling issues today. Thanks for those screen recordings!

**Comment by jasperhuangg** [comment]
To clarify, Tim opened a PR earlier that included fixes for this issue on iOS Native. So on Tim's PR, the `TermsAndLicenses` component successfully avoids the keyboard on the following platforms:
- iOS Native ✅ 
- Android Native ✅
- Mobile Chrome (login form ✅, working towards a fix for the password form)

We were unable to figure out a fix for Mobile Safari, and we have agreed to leave it for now. As long as the view is still scrollable and people can blur the text input, then the usability of the page is not really diminished.

**Comment by jasperhuangg** [comment]
For Mobile Chrome (on Android), the `TermsAndLicenses` component does actually shift up to avoid the keyboard. It turns out that there's just too many items occupying space on the screen for it to be shifted up enough for it to remain in view.

When I mean "too many items", compare these screenshots of the password form on the same Android emulator on both Android Native (right) and Mobile Chrome (left):

<p float="left">
  <img width="300" alt="Screen Shot 2022-12-13 at 4 43 56 PM" src="https://user-images.githubusercontent.com/31285285/207270075-e96e562e-16cd-4801-b704-e75518121b51.png">
  <img width="303" alt="Screen Shot 2022-12-13 at 4 43 44 PM" src="https://user-images.githubusercontent.com/31285285/207270087-5f41febe-9b31-417d-8066-c74af66ca681.png">
</p>

Notice how Mobile Chrome includes a nav bar and a password bar, all which occupy just enough height for the `TermsAndLicenses` component to get cut off by the keyboard.

We could conditionally reduce the margins/padding of everything on the screen to get it to fit, but since you can still scroll down to see the `TermsAndLicenses`, I think we should just do nothing here like we are for iOS Safari.

**Comment by jasperhuangg** [comment]
Totally agree! I definitely think we should keep the changes from your PR since we'll probably need them later on anyways when we add the footer. Thanks for keeping everyone on the same page and your help on this issue!

**Comment by tgolen** [comment]
I'm checking off most of those as NA since this has been an issue ever since the page was built

**Comment by JmillsExpensify** [comment]
Great thank you! I'll circle back on the regression test/payment once I return on Tuesday.

**Comment by JmillsExpensify** [comment]
Alright, circling back on this issue for the payment. It's due tomorrow, so let's make sure all is in order before then. @sobitneupane I've just invited you to the Upwork job. Mind accepting?

**Comment by JmillsExpensify** [comment]
@gadhiyamanan You're also eligible for the reporting bonus. Please apply to the [Upwork job](https://www.upwork.com/jobs/~0164d0286c2126643f).

**Comment by JmillsExpensify** [comment]
Thanks offer sent! 

Also I closed the loop on the regression test, so I'm crossing that off above in the BZ checklist.

**Comment by JmillsExpensify** [comment]
Summarizing how payments will shape out. 

 - External issue reporter: @gadhiyamanan $250
 - Contributor+ (internal PR review): $1,000

**Comment by JmillsExpensify** [comment]
Just realized that I didn't get around to paying this out before the Christmas holiday in the US, so I've just circled back and paid out both Contributors per the payments above. We already have a keyboard regression test is play, so I'm closing out this issue.
